processor/stream/stream_processor.go

```diff
@@ -227,9 +260,11 @@ func (s *StreamProcessor) HandleStream(ctx context.Context, meta map[string]inte
 		Config:      s.Tconfig,
 		Metadata:    *metadata,
 	}
+	rl := rateLimiterFromContext(ctx)
 
 	for {
-		transformables, done := s.readBatch(batchSize, jsonReader, res)
+
+		transformables, done := s.readBatch(ctx, rl, batchSize, jsonReader, res)
```

Comment:
You pointed out in a previous comment that we can fetch the `rate_limiter` once and not for every batch. I agree that this is more performant. The `context` is needed for the `WaitN` though. Thus I suggest to stick with the current solution for this.

Reply:
to be clear, I'm just suggesting to move

> rl := rateLimiterFromContext(ctx)

inside `readBatch` and removing the `rl` argument to `readBatch`. 

What we discussed was only accessing the LRU once, which would still be the case. I don't have strong feelings about this, but since the rate limiter is already in the context and we need to pass in the context anyway, i don't see the point of taking it out here and passing it in as a separate argument.

Reply:
I was referring to https://github.com/elastic/apm-server/pull/1367#discussion_r217454248: 
> I also think you should move:
limiter, limiterOk := ctx.Value(RateLimiterKey).(*rate.Limiter)
out of the loop. No reason to get it from the context every time.

This way we save a fetching and typecasting the limiter from the context for every batch.

Reply:
Me neither, but for clarification, _out of the loop_ in the above example translates to what I've done. If I move it to `readBatch` then it would be _inside the loop_ again.

Reply:
I don't think it's the same situation. The example I commented on previously requires that a line is moved up above the `for {` line - and nothing else. We were only using the rate limiter outside readBatch and not passing it in.

I think a good argument to continue passing in the rate limit is that it makes it explicit to `readBatch`. If we pass the rate limit in through the context only it's implicit and difficult for a caller to know that the context can contain a rate limit that `readBatch` will use.

Reply:
It is obvious that it is not completely the same situation as the code changed. I am not arguing for passing a parameter to a method or not. I am arguing for fetching and typecasting the rate limiter _once_ from the context and not for every batch.